app/code/Magento/Inventory/Model/SourceItem/Validator/StatusConsistencyValidator.php

```diff
@@ -0,0 +1,159 @@
+     * @param SourceItemStatus $sourceItemStatus
+     * @param StockConfigurationInterface $stockConfiguration
+     * @param LegacyStockItemRepository $legacyStockItemRepository
+     * @param ProductResourceModel $productResource
+     * @param StockItemCriteriaInterfaceFactory $stockItemCriteriaFactory
+     * @param IsValidationAllowed $isValidationAllowed
+     */
+    public function __construct(
+        ValidationResultFactory $validationResultFactory,
+        SourceItemStatus $sourceItemStatus,
+        StockConfigurationInterface $stockConfiguration,
+        LegacyStockItemRepository $legacyStockItemRepository,
+        ProductResourceModel $productResource,
+        StockItemCriteriaInterfaceFactory $stockItemCriteriaFactory,
+        IsValidationAllowed $isValidationAllowed
+    ) {
+        $this->validationResultFactory = $validationResultFactory;
+        $this->sourceItemStatus = $sourceItemStatus;
+        $this->stockConfiguration = $stockConfiguration;
+        $this->legacyStockItemRepository = $legacyStockItemRepository;
+        $this->productResource = $productResource;
+        $this->stockItemCriteriaFactory = $stockItemCriteriaFactory;
+        $this->isValidationAllowed = $isValidationAllowed;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function validate(SourceItemInterface $sourceItem): ValidationResult
+    {
+        $errors = [];
+        if ($this->isValidationAllowed->execute($sourceItem)) {
+            $quantity = $sourceItem->getQuantity();
+            $status = $sourceItem->getStatus();
+
+            $legacyStockItem = $this->getLegacyStockItem($sourceItem->getSku());
+            if (null === $legacyStockItem) {
+                $isManageStock = true;
+            } else {
+                $isManageStock = $this->isManageStock($legacyStockItem);
+            }
+
+            if ($isManageStock
+                && is_numeric($quantity)
+                && (float)$quantity <= 0
+                && (int)$status === SourceItemInterface::STATUS_IN_STOCK
+            ) {
+                $statusOptions = $this->sourceItemStatus->toOptionArray();
+                $labels = array_column($statusOptions, 'label', 'value');
+                $errors[] = __(
+                    'Product cannot have "%status" "%in_stock" while product "%quantity" equals or below zero',
+                    [
+                        'status' => SourceItemInterface::STATUS,
+                        'in_stock' => $labels[SourceItemInterface::STATUS_IN_STOCK],
+                        'quantity' => SourceItemInterface::QUANTITY
+                    ]
+                );
+            }
+        }
+
+        return $this->validationResultFactory->create(['errors' => $errors]);
+    }
+
+    /**
+     * @param string $sku
+     * @return LegacyStockItem|null
+     */
+    private function getLegacyStockItem(string $sku)
+    {
+        $productIds = $this->productResource->getProductsIdsBySkus([$sku]);
```

Comment:
You need to use `Magento\InventoryCatalog\Model\GetSkusByProductIdsInterface`

Reply:
This is interface to get sku by product id , but I need product id by sku.